File: backend/src/main/java/com/back/back9/domain/exchange/service/ExchangeService.java
package com.back.back9.domain.exchange.service;

import com.back.back9.domain.exchange.dto.CoinPriceResponse;
import com.back.back9.domain.exchange.dto.ExchangeDTO;
import com.back.back9.domain.websocket.mock.MockCoinListProvider;
import com.back.back9.domain.websocket.vo.CandleInterval;
import com.back.back9.global.redis.service.RedisService;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ExchangeService {

    private final RedisService redisService;
    private final ObjectMapper objectMapper;
    private final MockCoinListProvider coinListProvider;

    /**
     * 📌 Redis에서 초기 120개 캔들 가져오기
     */
    public List<ExchangeDTO> getInitialCandles(CandleInterval interval, String symbol) {
        List<JsonNode> jsonList = redisService.getLatestCandle(interval, symbol);
        return jsonList.stream()
                .map(json -> {
                    ExchangeDTO dto = objectMapper.convertValue(json, ExchangeDTO.class);
                    coinListProvider.getNameBySymbol(dto.getSymbol()).ifPresent(dto::setName);
                    return dto;
                })
                .collect(Collectors.toList());
    }

    /**
     * 📌 Redis에서 과거 캔들 페이지 단위로 가져오기 (스크롤)
     */
    public List<ExchangeDTO> getPreviousCandles(CandleInterval interval, String market, int page, LocalDateTime time) {
        List<JsonNode> jsonList = redisService.getPreviousCandlesByRange(interval, market, page, time);
        return jsonList.stream()
                .map(json -> {
                    ExchangeDTO dto = objectMapper.convertValue(json, ExchangeDTO.class);
                    coinListProvider.getNameBySymbol(dto.getSymbol()).ifPresent(dto::setName);
                    return dto;
                })
                .collect(Collectors.toList());
    }

    /**
     * 📌 Redis에서 1초봉 최신 캔들 1개 스캔 후 반환 (단일 심볼)
     */
    public CoinPriceResponse getLatestCandleByScan(String coinName) {
        try {
            String coinSymbol = CoinSymbolMapper.toMarketCode(coinName);
            JsonNode json = redisService.getLatest1sCandle(coinSymbol);

            if (json == null) {
                return new CoinPriceResponse(coinSymbol, BigDecimal.ZERO, LocalDateTime.now());
            }

            BigDecimal close = json.get("trade_price").decimalValue();
            LocalDateTime now = LocalDateTime.now().withNano(0);

            return new CoinPriceResponse(coinSymbol, close, now);
        } catch (Exception e) {
            throw new RuntimeException("getLatestCandleByScan 오류: " + e.getMessage(), e);
        }
    }

    /**
     * 📌 Redis에 저장된 최신 1초 캔들을 기반으로 코인 리스트 전체 조회
     * → MockCoinListProvider에서 한글명 매핑 포함
     */
    public List<ExchangeDTO> getCoinsLatest() {
        List<String> coinlist = coinListProvider.getMarketCodes();
        List<ExchangeDTO> coinInfoList = new ArrayList<>();

        for (String coinSymbol : coinlist) {
            JsonNode jsonNode = redisService.getLatest1sCandle(coinSymbol);
            if (jsonNode != null) {
                ExchangeDTO exchangeDTO = objectMapper.convertValue(jsonNode, ExchangeDTO.class);
                coinListProvider.getNameBySymbol(coinSymbol).ifPresent(exchangeDTO::setName);
                coinInfoList.add(exchangeDTO);
            }
        }

        return coinInfoList;
    }
}

File: backend/src/main/kotlin/com/back/back9/domain/exchange/service/ExchangeService.kt
package com.back.back9.domain.exchange.service

import com.back.back9.domain.exchange.dto.CandleResponseDTO
import com.back.back9.domain.exchange.dto.CoinPriceResponseDTO
import com.back.back9.domain.exchange.vo.CandleFetchParameters
import com.back.back9.domain.websocket.service.DatabaseCoinListProvider
import com.back.back9.domain.websocket.vo.CandleInterval
import com.back.back9.global.redis.dto.Candle
import com.back.back9.global.redis.service.RedisService
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

@Service
class ExchangeService(
    private val redisService: RedisService,
    private val coinListProvider: DatabaseCoinListProvider
) {

    fun getInitialCandles(interval: CandleInterval, symbol: String): List<CandleResponseDTO> {
        val key = interval.redisKey(symbol)
        val candles = redisService.getLatestCandles(key, CandleFetchParameters.INITIAL_CANDLE_COUNT)
        return candles.map { candle -> convertToCandleResponseDTO(candle, symbol) }
    }

    fun getPreviousCandles(interval: CandleInterval, market: String, cursorTimestamp: Long): List<CandleResponseDTO> {
        val key = interval.redisKey(market)
        val candles = redisService.getCandlesBefore(key, cursorTimestamp, CandleFetchParameters.PREVIOUS_CANDLE_COUNT)
        return candles.map { candle -> convertToCandleResponseDTO(candle, market) }
    }

    val coinsLatest: List<CandleResponseDTO>
        get() {
            return coinListProvider.getMarketCodes().mapNotNull { coinSymbol ->
                val key = CandleInterval.SEC.redisKey(coinSymbol)
                redisService.getLatestCandle(key)?.let { candle ->
                    val dto = convertToCandleResponseDTO(candle, coinSymbol)
                    // `apply` 블록 안에서는 `dto` 객체가 컨텍스트(this)가 되므로,
                    // 프로퍼티에 바로 접근할 수 있습니다.
                    dto.apply {
                        // [수정] CandleResponseDTO.setName이 아니라 인스턴스의 name 프로퍼티에 값을 할당합니다.
                        name = coinListProvider.getNameBySymbol(coinSymbol)
                    }
                }
            }
        }

    fun getLatestCandleByScan(coinName: String): CoinPriceResponseDTO {
        val key = CandleInterval.SEC.redisKey(coinName)
        return redisService.getLatestCandle(key)?.let { candle ->
            CoinPriceResponseDTO(
                coinName,
                candle.tradePrice,
                LocalDateTime.now().withNano(0)
            )
        } ?: CoinPriceResponseDTO(coinName, BigDecimal.ZERO, LocalDateTime.now())
    }

    private fun convertToCandleResponseDTO(candle: Candle, symbol: String): CandleResponseDTO {
        val dateTime = LocalDateTime.ofInstant(Instant.ofEpochMilli(candle.timestamp), ZoneId.systemDefault())
        return CandleResponseDTO(
            candle.timestamp, dateTime, symbol,
            candle.openingPrice, candle.highPrice, candle.lowPrice,
            candle.tradePrice, candle.candleAccTradeVolume
        )
    }
}